File: engine/static/input/src/controller.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/memory.hpp"
#include "argus/lowlevel/collections.hpp"

#include "argus/input/controller.hpp"
#include "argus/input/input_manager.hpp"
#include "argus/input/keyboard.hpp"
#include "argus/input/mouse.hpp"
#include "internal/input/controller.hpp"
#include "internal/input/gamepad.hpp"
#include "internal/input/pimpl/controller.hpp"
#include "internal/input/pimpl/input_manager.hpp"

#include <algorithm>
#include <string>
#include <vector>

namespace argus::input {
    static PoolAllocator g_pimpl_pool(sizeof(pimpl_Controller));

    double get_global_deadzone_radius(double radius);

    void set_global_deadzone_radius(double radius);

    DeadzoneShape get_global_deadzone_shape(void);

    void set_global_deadzone_shape(DeadzoneShape shape);

    Controller::Controller(const std::string &name):
        m_pimpl(&g_pimpl_pool.construct<pimpl_Controller>(name)) {
    }

    Controller::~Controller(void) {
        if (m_pimpl != nullptr) {
            g_pimpl_pool.destroy(m_pimpl);
        }
    }

    const std::string &Controller::get_name(void) const {
        return m_pimpl->name;
    }

    bool Controller::has_gamepad(void) const {
        return m_pimpl->attached_gamepad.has_value();
    }

    void Controller::attach_gamepad(HidDeviceId id) {
        if (m_pimpl->attached_gamepad.has_value()) {
            throw std::invalid_argument("Controller already has associated gamepad");
        }

        assoc_gamepad(id, this->get_name());
        m_pimpl->attached_gamepad = id;

        Logger::default_logger().info("Attached gamepad '%s' to controller '%s'",
                this->get_gamepad_name().c_str(), this->get_name().c_str());
    }

    bool Controller::attach_first_available_gamepad(void) {
        if (m_pimpl->attached_gamepad.has_value()) {
            throw std::invalid_argument("Controller already has associated gamepad");
        }

        auto id = assoc_first_available_gamepad(this->get_name());
        if (id < 0) {
            return false;
        }

        m_pimpl->attached_gamepad = id;

        Logger::default_logger().info("Attached gamepad '%s' to controller '%s'",
                this->get_gamepad_name().c_str(), this->get_name().c_str());

        return true;
    }

    void Controller::detach_gamepad(void) {
        if (!m_pimpl->attached_gamepad.has_value()) {
            // silently fail
            return;
        }

        unassoc_gamepad(m_pimpl->attached_gamepad.value());

        m_pimpl->attached_gamepad = std::nullopt;
    }

    std::string Controller::get_gamepad_name(void) {
        if (!this->has_gamepad()) {
            throw std::runtime_error("Controller does not have associated gamepad");
        }

        return ::argus::input::get_gamepad_name(m_pimpl->attached_gamepad.value());
    }

    double Controller::get_deadzone_radius(void) {
        return m_pimpl->dz_radius.value_or(InputManager::instance().get_global_deadzone_radius());
    }

    void Controller::set_deadzone_radius(double radius) {
        m_pimpl->dz_radius = std::min(std::max(radius, 0.0), 1.0);
    }

    void Controller::clear_deadzone_radius(void) {
        m_pimpl->dz_radius.reset();
    }

    DeadzoneShape Controller::get_deadzone_shape(void) {
        return m_pimpl->dz_shape.value_or(InputManager::instance().get_global_deadzone_shape());
    }

    void Controller::set_deadzone_shape(DeadzoneShape shape) {
        if (shape >= DeadzoneShape::MaxValue) {
            throw std::invalid_argument("Invalid deadzone shape ordinal "
                    + std::to_string(std::underlying_type_t<decltype(shape)>(shape)));
        }
        m_pimpl->dz_shape = shape;
    }

    void Controller::clear_deadzone_shape(void) {
        m_pimpl->dz_shape.reset();
    }

    static void _check_axis(GamepadAxis axis) {
        if (axis >= GamepadAxis::MaxValue) {
            throw std::invalid_argument("Invalid gamepad axis ordinal "
                    + std::to_string(std::underlying_type_t<decltype(axis)>(axis)));
        }
    }

    double Controller::get_axis_deadzone_radius(GamepadAxis axis) {
        _check_axis(axis);
        return m_pimpl->dz_axis_radii.at(size_t(axis)).value_or(
                m_pimpl->dz_radius.value_or(
                        InputManager::instance().get_global_axis_deadzone_radius(axis)));
    }

    void Controller::set_axis_deadzone_radius(GamepadAxis axis, double radius) {
        _check_axis(axis);
        m_pimpl->dz_axis_radii.at(size_t(axis)) = std::min(std::max(radius, 0.0), 1.0);
    }

    void Controller::clear_axis_deadzone_radius(GamepadAxis axis) {
        _check_axis(axis);
        m_pimpl->dz_axis_radii.at(size_t(axis)).reset();
    }

    DeadzoneShape Controller::get_axis_deadzone_shape(GamepadAxis axis) {
        _check_axis(axis);
        return m_pimpl->dz_axis_shapes.at(size_t(axis)).value_or(
                m_pimpl->dz_shape.value_or(
                        InputManager::instance().get_global_axis_deadzone_shape(axis)));
    }

    void Controller::set_axis_deadzone_shape(GamepadAxis axis, DeadzoneShape shape) {
        _check_axis(axis);
        m_pimpl->dz_axis_shapes.at(size_t(axis)) = shape;
    }

    void Controller::clear_axis_deadzone_shape(GamepadAxis axis) {
        _check_axis(axis);
        m_pimpl->dz_axis_shapes.at(size_t(axis)).reset();
    }

    void Controller::unbind_action(const std::string &action) {
        if (m_pimpl->action_to_key_bindings.find(action) == m_pimpl->action_to_key_bindings.end()) {
            return;
        }

        // remove action from binding list of keys it's bound to
        for (auto key : m_pimpl->action_to_key_bindings[action]) {
            remove_from_vector(m_pimpl->key_to_action_bindings[key], action);
        }

        // remove binding list of action
        m_pimpl->action_to_key_bindings.erase(action);
    }

    template<typename T>
    static void _bind_thing(std::map<T, std::vector<std::string>> &to_map,
            std::map<std::string, std::vector<T>> &from_map, T thing, const std::string action) {
        // we maintain two binding maps because actions and "things" have a
        // many-to-many relationship; i.e. each key may be bound to multiple
        // actions and each action may have multiple keys bound to it

        // insert into the thing-to-actions map
        auto &tta_vec = to_map[thing];
        if (std::find(tta_vec.begin(), tta_vec.end(), action) == tta_vec.end()) {
            tta_vec.push_back(action);
        }
        // nothing to do if the binding already exists

        // insert into the action-to-things map
        auto &att_vec = from_map[action];
        if (std::find(att_vec.begin(), att_vec.end(), thing) == att_vec.end()) {
            att_vec.push_back(thing);
        }
        // nothing to do if the binding already exists
    }

    template<typename T>
    static void _unbind_thing(std::map<T, std::vector<std::string>> &to_map,
            std::map<std::string, std::vector<T>> &from_map, T thing) {
        if (to_map.find(thing) == to_map.end()) {
            return;
        }

        // remove action from binding list of things it's bound to
        for (const auto &action : to_map[thing]) {
            remove_from_vector(from_map[action], thing);
        }

        // remove binding list of thing
        to_map.erase(thing);
    }

    template<typename T>
    static void _unbind_thing(std::map<T, std::vector<std::string>> &to_map,
            std::map<std::string, std::vector<T>> &from_map, T thing, const std::string &action) {
        auto actions_it = from_map.find(action);
        if (actions_it != from_map.end()) {
            remove_from_vector(actions_it->second, thing);
        }

        auto things_it = to_map.find(thing);
        if (things_it != to_map.end()) {
            remove_from_vector(things_it->second, action);
        }
    }

    std::vector<std::string> Controller::get_keyboard_key_bindings(KeyboardScancode key) const {
        auto it = m_pimpl->key_to_action_bindings.find(key);
        if (it != m_pimpl->key_to_action_bindings.end()) {
            return it->second; // implicitly deep-copied
        }

        // no bindings so just return empty vector
        return {};
    }

    std::vector<KeyboardScancode> Controller::get_keyboard_action_bindings(const std::string &action) const {
        auto it = m_pimpl->action_to_key_bindings.find(action);
        if (it != m_pimpl->action_to_key_bindings.end()) {
            return it->second; // implicitly deep-copied
        }

        // no bindings so just return empty vector
        return {};
    }

    void Controller::bind_keyboard_key(KeyboardScancode key, const std::string &action) {
        _bind_thing(m_pimpl->key_to_action_bindings, m_pimpl->action_to_key_bindings, key, action);
    }

    void Controller::unbind_keyboard_key(KeyboardScancode key) {
        _unbind_thing(m_pimpl->key_to_action_bindings, m_pimpl->action_to_key_bindings, key);
    }

    void Controller::unbind_keyboard_key(KeyboardScancode key, const std::string &action) {
        _unbind_thing(m_pimpl->key_to_action_bindings, m_pimpl->action_to_key_bindings, key, action);
    }

    void Controller::bind_mouse_button(MouseButton button, const std::string &action) {
        _bind_thing(m_pimpl->mouse_button_to_action_bindings, m_pimpl->action_to_mouse_button_bindings, button, action);
    }

    void Controller::unbind_mouse_button(MouseButton button) {
        _unbind_thing(m_pimpl->mouse_button_to_action_bindings, m_pimpl->action_to_mouse_button_bindings, button);
    }

    void Controller::unbind_mouse_button(MouseButton button, const std::string &action) {
        _unbind_thing(m_pimpl->mouse_button_to_action_bindings, m_pimpl->action_to_mouse_button_bindings, button,
                action);//TODO
    }

    void Controller::bind_mouse_axis(MouseAxis axis, const std::string &action) {
        _bind_thing(m_pimpl->mouse_axis_to_action_bindings, m_pimpl->action_to_mouse_axis_bindings, axis, action);
    }

    void Controller::unbind_mouse_axis(MouseAxis axis) {
        _unbind_thing(m_pimpl->mouse_axis_to_action_bindings, m_pimpl->action_to_mouse_axis_bindings, axis);
    }

    void Controller::unbind_mouse_axis(MouseAxis axis, const std::string &action) {
        _unbind_thing(m_pimpl->mouse_axis_to_action_bindings, m_pimpl->action_to_mouse_axis_bindings, axis, action);
    }

    void Controller::bind_gamepad_button(GamepadButton button, const std::string &action) {
        _bind_thing(m_pimpl->gamepad_button_to_action_bindings, m_pimpl->action_to_gamepad_button_bindings, button,
                action);
    }

    void Controller::unbind_gamepad_button(GamepadButton button) {
        _unbind_thing(m_pimpl->gamepad_button_to_action_bindings, m_pimpl->action_to_gamepad_button_bindings, button);
    }

    void Controller::unbind_gamepad_button(GamepadButton button, const std::string &action) {
        _unbind_thing(m_pimpl->gamepad_button_to_action_bindings, m_pimpl->action_to_gamepad_button_bindings,
                button, action);
    }

    void Controller::bind_gamepad_axis(GamepadAxis button, const std::string &action) {
        _bind_thing(m_pimpl->gamepad_axis_to_action_bindings, m_pimpl->action_to_gamepad_axis_bindings, button, action);
    }

    void Controller::unbind_gamepad_axis(GamepadAxis button) {
        _unbind_thing(m_pimpl->gamepad_axis_to_action_bindings, m_pimpl->action_to_gamepad_axis_bindings, button);
    }

    void Controller::unbind_gamepad_axis(GamepadAxis button, const std::string &action) {
        _unbind_thing(m_pimpl->gamepad_axis_to_action_bindings, m_pimpl->action_to_gamepad_axis_bindings,
                button, action);
    }

    bool Controller::is_gamepad_button_pressed(GamepadButton button) {
        if (!has_gamepad()) {
            throw std::runtime_error("Cannot query gamepad button state for controller: No gamepad is associated");
        }

        return ::argus::input::is_gamepad_button_pressed(m_pimpl->attached_gamepad.value(), button);
    }

    double Controller::get_gamepad_axis(GamepadAxis axis) {
        if (!has_gamepad()) {
            throw std::runtime_error("Cannot query gamepad axis state for controller: No gamepad is associated");
        }

        return ::argus::input::get_gamepad_axis(m_pimpl->attached_gamepad.value(), axis);
    }

    double Controller::get_gamepad_axis_delta(GamepadAxis axis) {
        if (!has_gamepad()) {
            throw std::runtime_error("Cannot query gamepad axis state for controller: No gamepad is associated");
        }

        return ::argus::input::get_gamepad_axis_delta(m_pimpl->attached_gamepad.value(), axis);
    }

    bool Controller::is_action_pressed(const std::string &action) {
        auto kb_it = m_pimpl->action_to_key_bindings.find(action);
        if (kb_it != m_pimpl->action_to_key_bindings.cend()) {
            for (auto key : kb_it->second) {
                if (is_key_pressed(key)) {
                    return true;
                }
            }
        }

        if (this->has_gamepad()) {
            auto gamepad_it = m_pimpl->action_to_gamepad_button_bindings.find(action);
            if (gamepad_it != m_pimpl->action_to_gamepad_button_bindings.cend()) {
                for (auto btn : gamepad_it->second) {
                    if (this->is_gamepad_button_pressed(btn)) {
                        return true;
                    }
                }
            }
        }

        auto mouse_it = m_pimpl->action_to_mouse_button_bindings.find(action);
        if (mouse_it != m_pimpl->action_to_mouse_button_bindings.cend()) {
            for (auto btn : mouse_it->second) {
                if (is_mouse_button_pressed(btn)) {
                    return true;
                }
            }
        }

        return false;
    }

    double Controller::get_action_axis(const std::string &action) {
        if (this->has_gamepad()) {
            auto gamepad_it = m_pimpl->action_to_gamepad_axis_bindings.find(action);
            if (gamepad_it != m_pimpl->action_to_gamepad_axis_bindings.cend() && !gamepad_it->second.empty()) {
                return this->get_gamepad_axis(gamepad_it->second.front());
            }
        }

        auto mouse_it = m_pimpl->action_to_mouse_axis_bindings.find(action);
        if (mouse_it != m_pimpl->action_to_mouse_axis_bindings.cend() && !mouse_it->second.empty()) {
            return get_mouse_axis(mouse_it->second.front());
        }

        return 0;
    }

    double Controller::get_action_axis_delta(const std::string &action) {
        if (this->has_gamepad()) {
            auto gamepad_it = m_pimpl->action_to_gamepad_axis_bindings.find(action);
            if (gamepad_it != m_pimpl->action_to_gamepad_axis_bindings.cend() && !gamepad_it->second.empty()) {
                return this->get_gamepad_axis_delta(gamepad_it->second.front());
            }
        }

        auto mouse_it = m_pimpl->action_to_mouse_axis_bindings.find(action);
        if (mouse_it != m_pimpl->action_to_mouse_axis_bindings.cend() && !mouse_it->second.empty()) {
            return get_mouse_axis_delta(mouse_it->second.front());
        }

        return 0;
    }

    void ack_gamepad_disconnects(void) {
        for (auto &[_, controller] : InputManager::instance().m_pimpl->controllers) {
            if (controller->m_pimpl->was_gamepad_disconnected) {
                // acknowledge disconnect flag set by render thread and fully
                // disassoc gamepad from controller
                controller->m_pimpl->was_gamepad_disconnected = false;
                controller->detach_gamepad();
            }
        }
    }
}

File: engine/modules/input/src/controller.rs
/*
* This file is a part of Argus.
* Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
use std::collections::{HashMap, HashSet};
use std::hash::Hash;
use argus_logging::info;
use argus_scripting_bind::script_bind;
use crate::{DeadzoneConfig, InputManager, LOGGER};
use crate::gamepad::*;
use crate::keyboard::{is_key_pressed, KeyboardScancode};
use crate::mouse::*;

#[script_bind(ref_only)]
pub struct Controller {
    name: String,
    attached_gamepad: Option<i32>,
    pub(crate) was_gamepad_disconnected: bool,

    deadzones: DeadzoneConfig,
    bindings: ControllerBindings,
}

#[derive(Default)]
struct ControllerBindings {
    key_to_actions: HashMap<KeyboardScancode, HashSet<String>>,
    action_to_keys: HashMap<String, HashSet<KeyboardScancode>>,

    mouse_button_to_actions: HashMap<MouseButton, HashSet<String>>,
    action_to_mouse_buttons: HashMap<String, HashSet<MouseButton>>,

    mouse_axis_to_actions: HashMap<MouseAxis, HashSet<String>>,
    action_to_mouse_axes: HashMap<String, HashSet<MouseAxis>>,

    gamepad_button_to_actions: HashMap<GamepadButton, HashSet<String>>,
    action_to_gamepad_buttons: HashMap<String, HashSet<GamepadButton>>,

    gamepad_axis_to_actions: HashMap<GamepadAxis, HashSet<String>>,
    action_to_gamepad_axes: HashMap<String, HashSet<GamepadAxis>>,
}

fn max_abs(a: f64, b: f64) -> f64 {
    let a_sign = if a > 0.0 { 1.0 } else { -1.0 };
    let b_sign = if a > 0.0 { 1.0 } else { -1.0 };
    if a.abs() > b.abs() {
        a * a_sign
    } else {
        b * b_sign
    }
}

#[script_bind]
impl Controller {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            attached_gamepad: None,
            was_gamepad_disconnected: false,
            deadzones: Default::default(),
            bindings: Default::default(),
        }
    }

    #[script_bind]
    pub fn get_name(&self) -> &str {
        self.name.as_str()
    }

    #[script_bind]
    pub fn has_gamepad(&self) -> bool {
        self.attached_gamepad.is_some()
    }

    #[script_bind]
    //TODO: id should be HidDeviceInstanceId
    pub fn attach_gamepad(&mut self, id: i32) {
        if self.attached_gamepad.is_some() {
            panic!("Controller already has associated gamepad");
        }

        //TODO: figure out if we need to propagate this error
        assoc_gamepad(id, self.get_name()).expect("Failed to associate gamepad");
        self.attached_gamepad = Some(id);

        info!(
            LOGGER,
            "Attached gamepad '{}' to controller '{}'",
            self.get_gamepad_name(),
            self.get_name());
    }

    #[script_bind]
    pub fn attach_first_available_gamepad(&mut self) -> bool {
        if self.attached_gamepad.is_some() {
            panic!("Controller already has associated gamepad");
        }

        let Ok(id) = assoc_first_available_gamepad(self.get_name()) else { return false; };

        self.attached_gamepad = Some(id);

        info!(
            LOGGER,
            "Attached gamepad '{}' to controller '{}'",
            self.get_gamepad_name(),
            self.get_name()
        );

        true
    }

    #[script_bind]
    pub fn detach_gamepad(&mut self) {
        if let Some(gamepad) = self.attached_gamepad {
            unassoc_gamepad(gamepad);
            self.attached_gamepad = None;
        }
        // else silently fail
    }

    #[script_bind]
    pub fn get_gamepad_name(&self) -> String {
        match self.attached_gamepad {
            Some(gamepad) => get_gamepad_name(gamepad),
            None => panic!("Controller does not have an associated gamepad"),
        }
    }

    #[script_bind]
    pub fn get_deadzone_radius(&self) -> f64 {
        self.deadzones.radius.unwrap_or(InputManager::instance().get_global_deadzone_radius())
    }

    #[script_bind]
    pub fn set_deadzone_radius(&mut self, radius: f64) {
        self.deadzones.radius = Some(radius.clamp(0.0, 1.0));
    }

    #[script_bind]
    pub fn clear_deadzone_radius(&mut self) {
        self.deadzones.radius = None;
    }

    #[script_bind]
    pub fn get_deadzone_shape(&self) -> DeadzoneShape {
        self.deadzones.shape.unwrap_or(InputManager::instance().get_global_deadzone_shape())
    }

    #[script_bind]
    pub fn set_deadzone_shape(&mut self, shape: DeadzoneShape) {
        self.deadzones.shape = Some(shape);
    }

    #[script_bind]
    pub fn clear_deadzone_shape(&mut self) {
        self.deadzones.shape = None;
    }

    #[script_bind]
    pub fn get_axis_deadzone_radius(&self, axis: GamepadAxis) -> f64 {
        self.deadzones.axis_radii.get(&axis).cloned()
            .or_else(|| self.deadzones.radius)
            .unwrap_or_else(|| InputManager::instance().get_global_axis_deadzone_radius(&axis))
    }

    #[script_bind]
    pub fn set_axis_deadzone_radius(&mut self, axis: GamepadAxis, radius: f64) {
        self.deadzones.axis_radii.insert(axis, radius.clamp(0.0, 1.0));
    }

    #[script_bind]
    pub fn clear_axis_deadzone_radius(&mut self, axis: GamepadAxis) {
        self.deadzones.axis_radii.remove(&axis);
    }

    #[script_bind]
    pub fn get_axis_deadzone_shape(&self, axis: GamepadAxis) -> DeadzoneShape {
        self.deadzones.axis_shapes.get(&axis).cloned()
            .or_else(|| self.deadzones.shape)
            .unwrap_or_else(|| InputManager::instance().get_global_axis_deadzone_shape(&axis))
    }

    #[script_bind]
    pub fn set_axis_deadzone_shape(&mut self, axis: GamepadAxis, shape: DeadzoneShape) {
        self.deadzones.axis_shapes.insert(axis, shape);
    }

    pub fn clear_axis_deadzone_shape(&mut self, axis: GamepadAxis) {
        self.deadzones.axis_shapes.remove(&axis);
    }

    #[script_bind]
    pub fn unbind_action(&mut self, action: &str) {
        let Some(bound_keys) = self.bindings.action_to_keys.get(action) else { return; };

        // remove action from binding list of keys it's bound to
        for key in bound_keys {
            self.bindings.key_to_actions.remove(&key);
        }

        // remove binding list of action
        self.bindings.action_to_keys.remove(action);
    }

    pub fn get_keyboard_key_bindings(&self, key: KeyboardScancode) -> Option<&HashSet<String>> {
        self.bindings.key_to_actions.get(&key)
    }

    pub fn get_keyboard_action_bindings(&self, action: impl AsRef<str>)
        -> Option<&HashSet<KeyboardScancode>> {
        self.bindings.action_to_keys.get(action.as_ref())
    }

    #[script_bind]
    pub fn bind_keyboard_key(&mut self, key: KeyboardScancode, action: &str) {
        Self::bind_thing(
            &mut self.bindings.key_to_actions,
            &mut self.bindings.action_to_keys,
            key,
            action,
        );
    }

    #[script_bind]
    pub fn unbind_keyboard_key(&mut self, key: KeyboardScancode) {
        Self::unbind_thing(
            &mut self.bindings.key_to_actions,
            &mut self.bindings.action_to_keys,
            key,
            None,
        );
    }

    #[script_bind]
    pub fn unbind_keyboard_key_action(&mut self, key: KeyboardScancode, action: &str) {
        Self::unbind_thing(
            &mut self.bindings.key_to_actions,
            &mut self.bindings.action_to_keys,
            key,
            Some(action.as_ref()),
        );
    }

    #[script_bind]
    pub fn bind_mouse_button(&mut self, button: MouseButton, action: &str) {
        Self::bind_thing(
            &mut self.bindings.mouse_button_to_actions,
            &mut self.bindings.action_to_mouse_buttons,
            button,
            action,
        );
    }

    #[script_bind]
    pub fn unbind_mouse_button(&mut self, button: MouseButton) {
        Self::unbind_thing(
            &mut self.bindings.mouse_button_to_actions,
            &mut self.bindings.action_to_mouse_buttons,
            button,
            None,
        );
    }

    #[script_bind]
    pub fn unbind_mouse_button_action(&mut self, button: MouseButton, action: &str) {
        Self::unbind_thing(
            &mut self.bindings.mouse_button_to_actions,
            &mut self.bindings.action_to_mouse_buttons,
            button,
            Some(action),
        );
    }

    #[script_bind]
    pub fn bind_mouse_axis(&mut self, axis: MouseAxis, action: &str) {
        Self::bind_thing(
            &mut self.bindings.mouse_axis_to_actions,
            &mut self.bindings.action_to_mouse_axes,
            axis,
            action,
        );
    }

    #[script_bind]
    pub fn unbind_mouse_axis(&mut self, axis: MouseAxis) {
        Self::unbind_thing(
            &mut self.bindings.mouse_axis_to_actions,
            &mut self.bindings.action_to_mouse_axes,
            axis,
            None,
        );
    }

    #[script_bind]
    pub fn unbind_mouse_axis_action(&mut self, axis: MouseAxis, action: &str) {
        Self::unbind_thing(
            &mut self.bindings.mouse_axis_to_actions,
            &mut self.bindings.action_to_mouse_axes,
            axis,
            Some(action.as_ref()),
        );
    }

    #[script_bind]
    pub fn bind_gamepad_button(&mut self, button: GamepadButton, action: &str) {
        Self::bind_thing(
            &mut self.bindings.gamepad_button_to_actions,
            &mut self.bindings.action_to_gamepad_buttons,
            button,
            action,
        );
    }

    #[script_bind]
    pub fn unbind_gamepad_button(&mut self, button: GamepadButton) {
        Self::unbind_thing(
            &mut self.bindings.gamepad_button_to_actions,
            &mut self.bindings.action_to_gamepad_buttons,
            button,
            None,
        );
    }

    #[script_bind]
    pub fn unbind_gamepad_button_action(
        &mut self,
        button: GamepadButton,
        action: &str
    ) {
        Self::unbind_thing(
            &mut self.bindings.gamepad_button_to_actions,
            &mut self.bindings.action_to_gamepad_buttons,
            button,
            Some(action.as_ref()),
        );
    }

    #[script_bind]
    pub fn bind_gamepad_axis(&mut self, axis: GamepadAxis, action: &str) {
        Self::bind_thing(
            &mut self.bindings.gamepad_axis_to_actions,
            &mut self.bindings.action_to_gamepad_axes,
            axis,
            action,
        );
    }

    #[script_bind]
    pub fn unbind_gamepad_axis(&mut self, axis: GamepadAxis) {
        Self::unbind_thing(
            &mut self.bindings.gamepad_axis_to_actions,
            &mut self.bindings.action_to_gamepad_axes,
            axis,
            None,
        );
    }

    #[script_bind]
    pub fn unbind_gamepad_axis_action(&mut self, axis: GamepadAxis, action: &str) {
        Self::unbind_thing(
            &mut self.bindings.gamepad_axis_to_actions,
            &mut self.bindings.action_to_gamepad_axes,
            axis,
            Some(action.as_ref()),
        );
    }

    #[script_bind]
    pub fn is_gamepad_button_pressed(&self, button: GamepadButton) -> bool {
        let Some(gamepad) = self.attached_gamepad else {
            panic!("Cannot query gamepad button state for controller: No gamepad is associated");
        };

        is_gamepad_button_pressed(gamepad, button)
    }

    pub fn get_gamepad_axis(&self, axis: &GamepadAxis) -> f64 {
        let Some(gamepad) = self.attached_gamepad else {
            panic!("Cannot query gamepad axis state for controller: No gamepad is associated");
        };

        get_gamepad_axis(gamepad, axis)
    }

    #[script_bind]
    pub fn get_gamepad_axis_delta(&self, axis: GamepadAxis) -> f64 {
        let Some(gamepad) = self.attached_gamepad else {
            panic!("Cannot query gamepad axis state for controller: No gamepad is associated");
        };

        get_gamepad_axis_delta(gamepad, axis)
    }

    #[script_bind]
    pub fn is_action_pressed(&self, action: &str) -> bool {
        if let Some(key_bindings) = self.bindings.action_to_keys.get(action) {
            for key in key_bindings {
                if is_key_pressed(*key) {
                    return true;
                }
            }
        }

        if self.has_gamepad() {
            if let Some(gamepad_bindings) = self.bindings.action_to_gamepad_buttons
                .get(action) {
                for btn in gamepad_bindings {
                    if self.is_gamepad_button_pressed(*btn) {
                        return true;
                    }
                }
            }
        }

        if let Some(mouse_bindings) = self.bindings.action_to_mouse_buttons.get(action) {
            for btn in mouse_bindings {
                if is_mouse_button_pressed(*btn) {
                    return true;
                }
            }

        }

        false
    }

    #[script_bind]
    pub fn get_action_axis(&self, action: &str) -> f64 {
        let mut max_value: f64 = 0.0;
        if self.has_gamepad() {
            if let Some(gamepad_bindings) =
                self.bindings.action_to_gamepad_axes.get(action) {
                max_value = max_abs(
                    max_value,
                    gamepad_bindings.iter()
                        .map(|binding| self.get_gamepad_axis(binding))
                        .reduce(max_abs)
                        .unwrap_or(0.0),
                );
            }
        }

        if let Some(mouse_bindings) = self.bindings.action_to_mouse_axes.get(action) {
            max_value = max_abs(
                max_value,
                mouse_bindings.iter()
                    .map(|axis| get_mouse_axis(axis))
                    .reduce(max_abs)
                    .unwrap_or(0.0),
            );
        }

        max_value
    }

    #[script_bind]
    pub fn get_action_axis_delta(&self, action: &str) -> f64 {
        let mut max_value: f64 = 0.0;
        if self.has_gamepad() {
            if let Some(gamepad_bindings) =
                self.bindings.action_to_gamepad_axes.get(action) {
                max_value = max_abs(
                    max_value,
                    gamepad_bindings.iter()
                        .map(|axis| self.get_gamepad_axis_delta(*axis))
                        .reduce(max_abs)
                        .unwrap_or(0.0),
                );
            }
        }

        if let Some(mouse_bindings) = self.bindings.action_to_mouse_axes.get(action) {
            max_value = max_abs(
                max_value,
                mouse_bindings.iter()
                    .map(|axis| get_mouse_axis_delta (*axis))
                    .reduce(max_abs)
                    .unwrap_or(0.0),
            );
        }

        max_value
    }

    fn bind_thing<T: Copy + Eq + Hash>(
        to_map: &mut HashMap<T, HashSet<String>>,
        from_map: &mut HashMap<String, HashSet<T>>,
        thing: T,
        action: impl AsRef<str>
    ) {
        // we maintain two binding maps because actions and "things" have a
        // many-to-many relationship; i.e. each key may be bound to multiple
        // actions and each action may have multiple keys bound to it

        // insert into the thing-to-actions map
        to_map.entry(thing).or_default().insert(action.as_ref().to_string());

        // insert into the action-to-things map
        from_map.entry(action.as_ref().to_string()).or_default().insert(thing);
    }

    fn unbind_thing<T: Copy + Eq + Hash>(
        to_map: &mut HashMap<T, HashSet<String>>,
        from_map: &mut HashMap<String, HashSet<T>>,
        thing: T,
        action_opt: Option<&str>,
    ) {
        match action_opt {
            Some(action) => {
                if let Some(set) = from_map.get_mut(action) {
                    set.remove(&thing);
                }
                if let Some(v) = to_map.get_mut(&thing) {
                    v.remove(action);
                }
            }
            None => {
                for (_, bindings) in from_map {
                    bindings.remove(&thing);
                }
                to_map.remove(&thing);
            }
        }
    }

    pub(crate) fn get_mouse_button_actions(&self, button: &MouseButton)
        -> Option<&HashSet<String>> {
        self.bindings.mouse_button_to_actions.get(button)
    }

    pub(crate) fn get_mouse_axis_actions(&self, axis: &MouseAxis)
        -> Option<&HashSet<String>> {
        self.bindings.mouse_axis_to_actions.get(axis)
    }

    pub(crate) fn get_gamepad_button_actions(&self, button: &GamepadButton)
        -> Option<&HashSet<String>> {
        self.bindings.gamepad_button_to_actions.get(button)
    }

    pub(crate) fn get_gamepad_axis_actions(&self, axis: &GamepadAxis)
        -> Option<&HashSet<String>> {
        self.bindings.gamepad_axis_to_actions.get(axis)
    }
}

pub(crate) fn ack_gamepad_disconnects() {
    for mut item in InputManager::instance().controllers.iter_mut() {
        let controller = item.value_mut();
        if controller.was_gamepad_disconnected {
            // acknowledge disconnect flag set by render thread and fully
            // disassociate gamepad from controller
            controller.was_gamepad_disconnected = false;
            controller.detach_gamepad();
        }
    }
}
